third_party/amd/lib/TritonAMDGPUTransforms/inThreadTranspose.cpp

```diff
@@ -0,0 +1,210 @@
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+#define GEN_PASS_CLASSES
+#include "TritonAMDGPUTransforms/Passes.h.inc"
+
+using namespace mlir;
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+
+static Type getNewType(Type type, Attribute encoding) {
+  RankedTensorType tensorType = dyn_cast<RankedTensorType>(type);
+  return RankedTensorType::get(tensorType.getShape(),
+                               tensorType.getElementType(), encoding);
+}
+
+void convertLayout(Attribute encoding, Operation *op) {
+  OpBuilder builder(op);
+  // Convert operands
+  // For load/store with tensor pointers, we don't have to change the
+  // operands' type, we do this by changing the outputs' type of
+  // `make_tensor_ptr`
+  SmallVector<Value, 4> newArgs;
+  for (auto operand : op->getOperands()) {
+    auto tensorType = dyn_cast<RankedTensorType>(operand.getType());
+    if (tensorType &&
+        !isa<triton::gpu::SharedEncodingAttr>(tensorType.getEncoding())) {
+      Type newType = getNewType(tensorType, encoding);
+      newArgs.push_back(builder.create<triton::gpu::ConvertLayoutOp>(
+          op->getLoc(), newType, operand));
+    } else {
+      newArgs.push_back(operand);
+    }
+  }
+
+  // Convert output types
+  SmallVector<Type, 4> newTypes;
+  for (auto t : op->getResultTypes()) {
+    newTypes.push_back(getNewType(t, encoding));
+  }
+
+  // Construct new op with the new encoding
+  Operation *newOp = builder.create(op->getLoc(), op->getName().getIdentifier(),
+                                    newArgs, newTypes, op->getAttrs());
+
+  // Cast the results back to the original layout
+  for (size_t i = 0; i < op->getNumResults(); i++) {
+    Value newResult = newOp->getResult(i);
+    if (newTypes[i] != op->getResultTypes()[i]) {
+      newResult = builder.create<triton::gpu::ConvertLayoutOp>(
+          op->getLoc(), op->getResult(i).getType(), newResult);
+    }
+    op->getResult(i).replaceAllUsesWith(newResult);
+  }
+  op->erase();
+}
+
+SmallVector<Operation *> getLoadInsts(Operation *op) {
+  SmallVector<Operation *> ret;
+  // Deal with the case that convert_layout intakes a scf.if, etc.
+  auto v = op->getOperand(0);
+  auto prevOp = v.getDefiningOp();
+  if (isa<RegionBranchOpInterface>(prevOp)) {
+    LDBG("Dealing with scf blocks");
+    auto idx = cast<OpResult>(v).getResultNumber();
+    llvm::SmallVector<scf::YieldOp> yieldOps;
+    prevOp->walk([&](Operation *op) {
+      if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
+        yieldOps.push_back(yieldOp);
+      }
+    });
+
+    for (auto yieldOp : yieldOps) {
+      auto maybeLoadOp = yieldOp.getOperand(idx).getDefiningOp();
+      if (isa<tt::LoadOp>(maybeLoadOp))
+        ret.push_back(maybeLoadOp);
+    }
+  } else {
+    // regular case
+    LDBG("Regular cases");
+    if (isa<tt::LoadOp>(prevOp)) {
+      ret.push_back(prevOp);
+    } else {
+      LDBG("we assume load->convert_layout->dot chain but we cannot find it.");
+    }
+  }
```

Comment:
For this branch, I'd suggest you do
if (regionBranchOp) {
} else if (loadOp) {
} else {
  
}